### PR TITLE
fix `RowWriter`  index out of bounds error

### LINKED ISSUE #2910: index out of range error from datafusion_row::write::write_field
**Describe the bug**
index out of range error coming from datafusion_row::write::write_field 

**To Reproduce**
It happened when I ran a query against a proprietary data set.
SQL is,
```sql
SELECT 
    tag(column_1),
    COUNT(column_1)
FROM 
    csv
GROUP BY
    column_1
```
I haven't been able to reproduce it without that data set.

**Expected behavior**
It does not panic

**Additional context**
cargo 1.62.0 (a748cf5a3 2022-06-08)

```
thread 'tokio-runtime-worker' panicked at 'range end index 153 out of range for slice of length 152', library/core/src/slice/index.rs:73:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
   2: core::slice::index::slice_end_index_len_fail_rt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/slice/index.rs:73:5
   3: core::ops::function::FnOnce::call_once
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:248:5
   4: core::intrinsics::const_eval_select
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/intrinsics.rs:2372:5
   5: core::slice::index::slice_end_index_len_fail
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/slice/index.rs:67:9
   6: datafusion_row::writer::write_field
   7: datafusion_row::writer::write_row
   8: <datafusion::physical_plan::aggregates::row_hash::GroupedHashAggregateStreamV2 as futures_core::stream::Stream>::poll_next
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  10: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  11: tokio::runtime::task::harness::Harness<T,S>::poll
  12: std::thread::local::LocalKey<T>::with
  13: tokio::runtime::thread_pool::worker::Context::run_task
  14: tokio::runtime::thread_pool::worker::Context::run
  15: tokio::macros::scoped_tls::ScopedKey<T>::set
  16: tokio::runtime::thread_pool::worker::run
  17: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
  18: tokio::runtime::task::harness::Harness<T,S>::poll
  19: tokio::runtime::blocking::pool::Inner::run
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ArrowError(ExternalError(Execution("Join Error: task 15 panicked")))', src/main.rs:47:46
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
   2: core::result::unwrap_failed
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/result.rs:1785:5
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   4: std::thread::local::LocalKey<T>::with
   5: tokio::park::thread::CachedParkThread::block_on
   6: tokio::runtime::Runtime::block_on
   7: my_app::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'tokio-runtime-worker' panicked at 'range end index 128 out of range for slice of length 120', library/core/src/slice/index.rs:73:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
   2: core::slice::index::slice_end_index_len_fail_rt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/slice/index.rs:73:5
   3: core::ops::function::FnOnce::call_once
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:248:5
   4: core::intrinsics::const_eval_select
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/intrinsics.rs:2372:5
   5: core::slice::index::slice_end_index_len_fail
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/slice/index.rs:67:9
   6: datafusion_row::writer::write_field
   7: datafusion_row::writer::write_row
   8: <datafusion::physical_plan::aggregates::row_hash::GroupedHashAggregateStreamV2 as futures_core::stream::Stream>::poll_next
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  10: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  11: tokio::runtime::task::harness::Harness<T,S>::poll
  12: std::thread::local::LocalKey<T>::with
  13: tokio::runtime::thread_pool::worker::Context::run_task
  14: tokio::runtime::thread_pool::worker::Context::run
  15: tokio::macros::scoped_tls::ScopedKey<T>::set
  16: tokio::runtime::thread_pool::worker::run
  17: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
  18: tokio::runtime::task::harness::Harness<T,S>::poll
  19: tokio::runtime::blocking::pool::Inner::run
```

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #2910 #2963 .

 # Rationale for this change

The weird bug was reported, it can be reproduced locally in `row_writer_resize_test`, the reason being when the column is not nullable then 1 extra byte added to initial offset, and this breaks the buffer resize logic, causing out of range issues on `        self.data[varlena_offset..varlena_offset + size].copy_from_slice(bytes);
`

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by comphead** [comment]
@yjshen please check this PR draft, as I noticed the row_writer logic is mostly done by you

**Comment by thomas-k-cameron** [comment]
Hi,

I'm the one reported the bug.

If it is necessary for the resizing logic to decide on the basis of length, not the capacity, of the `data` field, then shouldn't we do the same for the `pub(crate) fn write_field_binary` as well?

I wonder if `pub(crate) fn write_field_binary`  could create a similar bug.

File: datafusion/row/src/writer.rs [hunk 1]
```diff
@@ -349,7 +349,7 @@ pub(crate) fn write_field_utf8(
     let from = from.as_any().downcast_ref::<StringArray>().unwrap();
     let s = from.value(row_idx);
     let new_width = to.current_width() + s.as_bytes().len();
-    if new_width > to.data.capacity() {
+    if new_width > to.data.len() {
```
Comment (liukun4515) [inline datafusion/row/src/writer.rs:352]: Do we need to change the  `write_field_utf8` like this?

**Comment by liukun4515** [comment]
@yjshen PTAL

**Comment by liukun4515** [comment]
@comphead  Thanks for your contribution.
But I'm confuse about the method to fix this issue.

For binary and utf8, we need to write the `offset and size` before the data, why not update the `new_width` using `to.current_width() + s.as_bytes().len() + 8`.
The 8 is the size of `offset and size` which is encoded as a u64.

File: datafusion/row/src/writer.rs [hunk 2]
```diff
@@ -349,7 +349,7 @@ pub(crate) fn write_field_utf8(
     let from = from.as_any().downcast_ref::<StringArray>().unwrap();
     let s = from.value(row_idx);
     let new_width = to.current_width() + s.as_bytes().len();
```
Comment (liukun4515) [inline datafusion/row/src/writer.rs:351]: Here, we can use `let new_width = to.current_width() + s.as_bytes().len() + 8`

Comment (liukun4515) [reply datafusion/row/src/writer.rs:351]: And in the `set_utf8` or `set_binary`, we can use 

```
        self.varlena_offset += (size + 8);
        self.varlena_width += (size +8);
```
to update the offset and width.

**Comment by yjshen** [comment]
I'm aware of this PR but ran out of time today. Will come back tomorrow morning.

File: datafusion/row/src/writer.rs (hunk 2 above)
Comment (comphead) [reply datafusion/row/src/writer.rs:351]: > And in the `set_utf8` or `set_binary`, we can use
> 
> ```
>         self.varlena_offset += (size + 8);
>         self.varlena_width += (size +8);
> ```
> 
> to update the offset and width.

That was also my vision. But my concern here if we add 8 extra bytes per field, it may bloat the row size significantly?

File: datafusion/core/src/dataframe.rs [hunk 3]
```diff
@@ -1266,4 +1266,45 @@ mod tests {
...
+        SELECT 
+            COUNT(1)
+        FROM 
+            test
+        GROUP BY
+            column_1"#;
+
+        // let df = ctx.sql("SELECT * FROM test").await.unwrap();
+        // df.show_limit(10).await.unwrap();
+        // dbg!(df.schema());
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:1303]: Should we leave it in?

**Review by alamb — APPROVED** [review]
Thanks @comphead 

I also think @yjshen  should review this if he has time before we merge this in

**Comment by comphead** [comment]
> Thanks @comphead
> 
> I also think @yjshen should review this if he has time before we merge this in

Hi @alamb I'm waiting for @yjshen and after his review I can do the same for binary datatype

File: datafusion/row/src/writer.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/row/src/writer.rs:351]: @comphead you can go through the code of `set_utf8`.
```
    fn set_utf8(&mut self, idx: usize, value: &str) {
        self.assert_index_valid(idx);
        let bytes = value.as_bytes();
        let size = bytes.len();
        // this line: write the offset and size to the row_buffer
        self.set_offset_size(idx, size as u32);
        let varlena_offset = self.varlena_offset;
        self.data[varlena_offset..varlena_offset + size].copy_from_slice(bytes);
        self.varlena_offset += size;
        self.varlena_width += size;
    }
```
Not per field, and variable length data type like utf8 and binary.
From the @yjshen  codebase, I think the layout of utf8 and binary should be
```
─ ─ ─ ─ ─ ─  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
│ offset and size│     utf8 data or binary data │
─ ─ ─ ─ ─ ─  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
```

File: datafusion/row/src/writer.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/row/src/writer.rs:352]: In my opinion, this fix is not right.
The `data` is used to buffer fixed size of layout and row data, if the new width or new length is less than the capacity of data, why need to expand the space.

File: datafusion/row/src/writer.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/row/src/writer.rs:351]: I figure out the corner case to panic for your fix.
You can add the test with string data type, but the each value of string empty, just like that
```
arrow::array::StringArray::from(vec!["","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""]);
```

Comment (yjshen) [reply datafusion/row/src/writer.rs:351]: @liukun4515 can elaborate on why there are two '+ 8' suggested?

```
    self.varlena_offset += (size + 8);
    self.varlena_width += (size +8);
```

File: datafusion/row/src/writer.rs (hunk 1 above)
Comment (yjshen) [reply datafusion/row/src/writer.rs:352]: >  if the new width or new length is less than the capacity of data, why need to expand the space.

I think it's because `Vec` in rust doesn't own the space between (len, capacity].

File: datafusion/row/src/writer.rs [hunk 4]
```diff
@@ -269,29 +269,29 @@ impl RowWriter {
 
 /// Stitch attributes of tuple in `batch` at `row_idx` and returns the tuple width
 pub fn write_row(
-    row: &mut RowWriter,
+    row_writer: &mut RowWriter,
```
Comment (yjshen) [inline datafusion/row/src/writer.rs:272]: This renaming is unrelated, I think?

File: datafusion/core/src/dataframe.rs (hunk 3 above)
Comment (yjshen) [reply datafusion/core/src/dataframe.rs:1303]: We'd better remove this I suppose.

**Review by yjshen — COMMENTED** [review]
Thanks @comphead!

I propose we fix the followings:

- The `new_width > to.data.len()` change makes sense to me, but on the resize part:
```rs
    if new_width > to.data.len() {
        to.data.resize(max(to.data.capacity(), new_width), 0);
    }
```
I think we could just resize to capacity to avoid reallocating, instead of the previous double capacity way.

- `write_field_binary` should be adapted accordingly as well.

- The capacity() bug also exists in the `end_padding` logic, it should be:

```rs
    /// End each row at 8-byte word boundary.
    pub(crate) fn end_padding(&mut self) {
        let payload_width = self.current_width();
        self.row_width = round_upto_power_of_2(payload_width, 8);
        if self.data.len() < self.row_width {
            self.data.resize(self.row_width, 0);
        }
    }
```

@liukun4515 I cannot quite get your `+8` logic while calculating size in the comments above, I didn't get panic with your 110 empty string case either. Please correct me if I have missed something important.

File: datafusion/row/src/writer.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/row/src/writer.rs:352]: thanks for your explain.
Please ignore my comments. @comphead

File: datafusion/row/src/writer.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/row/src/writer.rs:351]: Please ignore my comment, I have a error thought about the row layout when writing the comments

**Comment by liukun4515** [comment]
> Thanks @comphead!
> 
> I propose we fix the followings:
> 
> * The `new_width > to.data.len()` change makes sense to me, but on the resize part:
> 
> ```rust
>     if new_width > to.data.len() {
>         to.data.resize(max(to.data.capacity(), new_width), 0);
>     }
> ```
> 
> I think we could just resize to capacity to avoid reallocating, instead of the previous double capacity way.
> 
> * `write_field_binary` should be adapted accordingly as well.
> * The capacity() bug also exists in the `end_padding` logic, it should be:
> 
> ```rust
>     /// End each row at 8-byte word boundary.
>     pub(crate) fn end_padding(&mut self) {
>         let payload_width = self.current_width();
>         self.row_width = round_upto_power_of_2(payload_width, 8);
>         if self.data.len() < self.row_width {
>             self.data.resize(self.row_width, 0);
>         }
>     }
> ```
> 
> @liukun4515 I cannot quite get your `+8` logic while calculating size in the comments above, I didn't get panic with your 110 empty string case either. Please correct me if I have missed something important.

Sorry for the error comments with my wrong understanding for row layout.

File: datafusion/row/src/writer.rs (hunk 4 above)
Comment (comphead) [reply datafusion/row/src/writer.rs:272]: I thought its better naming, because when reading the code I supposed the `row`  is either incoming row or row buffer, both  are not correct. row_writer better reflects the object imho.

File: datafusion/row/src/writer.rs [hunk 5]
```diff
@@ -349,9 +349,9 @@ pub(crate) fn write_field_utf8(
     let from = from.as_any().downcast_ref::<StringArray>().unwrap();
     let s = from.value(row_idx);
     let new_width = to.current_width() + s.as_bytes().len();
-    if new_width > to.data.capacity() {
+    if new_width > to.data.len() {
         // double the capacity to avoid repeated resize
```
Comment (yjshen) [inline datafusion/row/src/writer.rs:353]: outdated code comments

File: datafusion/row/src/writer.rs [hunk 6]
```diff
@@ -365,9 +365,9 @@ pub(crate) fn write_field_binary(
     let from = from.as_any().downcast_ref::<BinaryArray>().unwrap();
     let s = from.value(row_idx);
     let new_width = to.current_width() + s.len();
-    if new_width > to.data.capacity() {
+    if new_width > to.data.len() {
         // double the capacity to avoid repeated resize
```
Comment (yjshen) [inline datafusion/row/src/writer.rs:369]: same here

**Review by yjshen — APPROVED** [review]

**Comment by yjshen** [comment]
@alamb wants to take another look?

**Review by alamb — APPROVED** [review]
looks great -- thank you all for your help and attention to this matter
